REMEMBER: FIRST CREATE A COPY OF THIS FILE WITH A UNIQUE NAME AND DO YOUR WORK THERE. AND MAKE SURE YOU COMMIT YOUR CHANGES TO THE `hw3_submissions` BRANCH.

# Assignment 3 | Cleaning and Exploring Data with Pandas



<img src="data/scoreCard.jpg" width=250>

In this assignment, you will investigate restaurant food safety scores for restaurants in San Francisco. Above is a sample score card for a restaurant. The scores and violation information have been made available by the San Francisco Department of Public Health. 

## Loading Food Safety Data


There are 2 files in the data directory:
1. business.csv containing food establishments in San Francisco
1. inspections.csv containing retaurant inspections records

Let's start by loading them into Pandas dataframes.  One of the files, business.csv, has encoding (ISO-8859-1), so you will need to account for that when reading it.

### Question 1

#### Question 1a
Read the two files noted above into two pandas dataframes named `bus` and `ins`, respectively. Print the first 5 rows of each to inspect them.


In [1]:
import pandas as pd

In [2]:
bus = pd.read_csv('data/businesses.csv', encoding='ISO-8859-1')
ins = pd.read_csv('data/inspections.csv')

In [3]:
bus.head(5)

business_id                                name  \
0           10                    TIRAMISU KITCHEN   
1           19               NRGIZE LIFESTYLE CAFE   
2           24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3           31      NORMAN'S ICE CREAM AND FREEZES   
4           45                 CHARLIE'S DELI CAFE   

                         address           city state postal_code   latitude  \
0                  033 BELDEN PL  San Francisco    CA       94104  37.791116   
1   1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109  37.786848   
2  500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104  37.792888   
3           2801 LEAVENWORTH ST   San Francisco    CA       94133  37.807155   
4                3202 FOLSOM ST   San Francisco    CA       94110  37.747114   

    longitude  phone_number  
0 -122.403816  +14154217044  
1 -122.421547  +14157763262  
2 -122.403135  +14156779494  
3 -122.419004           NaN  
4 -122.413641  +14156415051

In [4]:
ins.head(5)

business_id  score      date     type
0           10     82  20160503  routine
1           10     94  20140729  routine
2           10     92  20140114  routine
3           19     94  20160513  routine
4           19     94  20141110  routine

## Examining the Business data

From its name alone, we expect the `businesses.csv` file to contain information about the restaurants. Let's investigate this dataset.

### Question 2

#### Question 2a: How many records are there?

In [5]:
len(bus)

6315

In [6]:
bus.shape[0] #alternatively wecan use this where 0 specifies the starting index

6315

#### Question 2b: How many unique business IDs are there?  

In [7]:
bus.groupby ('business_id').nunique()

name  address  city  state  postal_code  latitude  longitude  \
business_id                                                                 
10              1        1     1      1            1         1          1   
19              1        1     1      1            1         1          1   
24              1        1     1      1            1         1          1   
31              1        1     1      1            1         1          1   
45              1        1     1      1            1         1          1   
...           ...      ...   ...    ...          ...       ...        ...   
90234           1        1     1      1            1         0          0   
90236           1        1     1      1            1         0          0   
90237           1        1     1      1            1         0          0   
90238           1        1     1      1            1         0          0   
90239           1        1     1      1            1         0          0   

             phone_number  
business_id                
10                      1  
19                      1  
24                      1  
31                      0  
45                      1  
...                   ...  
90234                   0  
90236                   1  
90237                   1  
90238                   1  
90239                   1  

[6315 rows x 8 columns]

In [8]:
print ('There are 6315 unique business IDs')

There are 6315 unique business IDs


#### Question 2c: What are the 5 most common businesses by name, and how many are there in San Francisco?

In [9]:
bus['name'].value_counts().head(5)

name
STARBUCKS COFFEE              72
PEET'S COFFEE & TEA           24
MCDONALDS                     12
SAN FRANCISCO SOUP COMPANY    11
WALGREENS                     11
Name: count, dtype: int64

## Zip code

Next, let's  explore some of the variables in the business table. We begin by examining the postal code.

### Question 3

#### Question 3a
How are the zip code values stored in python (i.e. data type)?

To answer this you might want to examine a particular entry.

In [10]:
type(bus.loc[0, 'postal_code']) #need to get a slice of the df, so we specify row index and column index

str

In [11]:
print('Zip code values are stored as Strings')

Zip code values are stored as Strings


#### Question 3b

What are the unique values of postal_code?

In [12]:
bus.groupby ('postal_code').nunique()

business_id  name  address  city  state  latitude  longitude  \
postal_code                                                                 
00000                  1     1        1     1      1         1          1   
92672                  1     1        1     1      1         1          1   
94013                  2     2        2     1      1         1          1   
94014                  2     2        2     1      1         0          0   
94066                  1     1        1     1      1         0          0   
941                    1     1        1     1      1         1          1   
94101                  2     2        2     1      1         1          1   
94102                458   434      418     1      1       211        211   
94103                558   527      458     1      1       201        204   
941033148              1     1        1     1      1         0          0   
94104                133   122      120     1      1        57         55   
94105                226   196      193     1      1        89         90   
94107                451   428      293     1      1       146        147   
94108                224   218      213     1      1       123        120   
94109                383   380      377     1      1       221        215   
94110                585   577      532     1      1       314        316   
941102019              1     1        1     1      1         1          1   
94111                282   256      223     1      1       108        110   
94112                190   189      188     1      1       126        126   
94114                230   223      209     1      1       110        111   
94115                238   236      228     1      1       138        136   
94116                 93    92       91     1      1        54         57   
94117                197   195      188     1      1       115        118   
94118                243   241      224     1      1       124        133   
94120                  1     1        1     1      1         1          1   
94121                161   160      159     1      1        98        104   
94122                270   265      260     1      1       116        120   
94123                169   166      168     1      1       109        111   
94124                164   161      147     1      1        74         74   
94127                 71    71       68     1      1        42         42   
94129                  2     2        2     1      1         0          0   
94130                  8     7        8     1      1         2          2   
94131                 48    48       47     1      1        31         32   
94132                134   132      100     1      1        34         34   
94133                417   414      387     1      1       208        207   
94134                 74    74       74     1      1        44         45   
94143                  5     5        4     1      1         1          1   
94158                 17    17        9     1      1         2          2   
94188                  1     1        1     1      1         0          0   
94545                  1     1        1     1      1         0          0   
94602                  1     1        1     1      1         0          0   
94609                  1     1        1     1      1         1          1   
94621                  1     1        1     1      1         0          0   
95105                  1     1        1     1      1         1          1   
CA                     3     3        3     1      1         2          2   
Ca                     3     3        3     1      1         0          0   

             phone_number  
postal_code                
00000                   1  
92672                   1  
94013                   1  
94014                   2  
94066                   1  
941                     1  
94101                   2  
94102                 366  
94103                 452  
94103314

#### Question 3c

Let's say we decide to exclude the businesses that have no zipcode for our analysis (which might include food trucks for example).  Use the list of valid 5-digit zip codes below to create a new dataframe called bus_valid, with only businesses whose postal_codes show up in this list of valid zipcodes. How many businesses are there in this new dataframe?

In [13]:
validZip = ["94102", "94103", "94104", "94105", "94107", "94108",
            "94109", "94110", "94111", "94112", "94114", "94115",
            "94116", "94117", "94118", "94121", "94122", "94123", 
            "94124", "94127", "94131", "94132", "94133", "94134"]

In [14]:
bus_valid = bus[bus['postal_code'].isin(validZip)] #here bus_valid is the new dataframe and we are checking to see if an object part of bus (i.e postal code) in validzip
len(bus_valid)

5999

## Latitude and Longitude

Another aspect of the data we want to consider is the prevalence of missing values. If many records have missing values then we might be concerned about whether the nonmissing values are representative of the population.

### Question 4
 
Consider the longitude and latitude in the business DataFrame. 

#### Question 4a

How many businesses are missing longitude values, working with only the businesses that are in the list of valid zipcodes?

In [15]:
len(bus_valid[pd.isnull(bus_valid['longitude'])])

2483

#### Question 4b

Create a new dataframe with one row for each valid zipcode. The dataframe should include the following three columns:

1. `postal_code`: Contains the zip codes in the `validZip` variable above.
2. `null_lon`: The number of businesses in that zipcode with missing `longitude` values.
3. `not_null_lon`: The number of businesses without missing `longitude` values.

In [16]:
data = []
for zip in validZip:
    rowsForZip = bus_valid[bus_valid['postal_code']==zip]
    null_lon = rowsForZip[pd.isnull(rowsForZip['longitude'])].shape[0]
    not_null_lon = rowsForZip[pd.notnull(rowsForZip['longitude'])].shape[0]
    data.append([zip, null_lon, not_null_lon])
    
valid = pd.DataFrame(data) #new df
valid.columns = ['postal_code', 'null_lon', 'not_null_lon']
valid

postal_code  null_lon  not_null_lon
0        94102       196           262
1        94103       252           306
2        94104        59            74
3        94105       100           126
4        94107       247           204
5        94108        82           142
6        94109       146           237
7        94110       251           334
8        94111       115           167
9        94112        59           131
10       94114       108           122
11       94115        76           162
12       94116        35            58
13       94117        67           130
14       94118       100           143
15       94121        55           106
16       94122       123           147
17       94123        53           116
18       94124        88            76
19       94127        28            43
20       94131        14            34
21       94132        65            69
22       94133       135           282
23       94134        29            45

#### 4c. Do any zip codes appear to have more than their 'fair share' of missing longitude? 

To answer this, you will want to compute the proportion of missing longitude values for each zip code, and print the proportion missing longitude, and print the top five zipcodes in descending order of proportion missing postal_code.


In [17]:
valid['proportion'] = (valid['null_lon'] / (valid['null_lon'] + valid['not_null_lon']))
valid.sort_values (by =['proportion'], ascending=False).head(5)
#% = x/x+y

postal_code  null_lon  not_null_lon  proportion
4        94107       247           204    0.547672
18       94124        88            76    0.536585
21       94132        65            69    0.485075
10       94114       108           122    0.469565
16       94122       123           147    0.455556

# Investigate the inspection data

Let's now turn to the inspection DataFrame. Earlier, we found that `ins` has 4 columns, these are named `business_id`, `score`, `date` and `type`.  In this section, we determine the granularity of `ins` and investigate the kinds of information provided for the inspections. 

### Question 5

#### Question 5a
As with the business data, assess whether there is one inspection record for each business, by counting how many rows are in the data and how many unique businesses there are in the data. If they are exactly the same number, it means there is only one inspection per business, clearly.

In [18]:
total_rows = len(ins)
print(total_rows)
unique_businesses = ins['business_id'].nunique()
print(unique_businesses)

if total_rows == unique_businesses:
    print("There is exactly one inspection record per business.")
else:
    print("There are multiple inspection records per business.")

15430
5730
There are multiple inspection records per business.


#### Question 5b

What values does `type` take on? How many occurrences of each value is in the DataFrame? Create a new dataframe named `ins2` by copying `ins` and keeping only records with values of `type` that occur more than 10 times in the original table. In other words, eliminate records that have values of `type` that occur rarely (< 10 times). Check the result to make sure rare types are eliminated.

In [37]:
ins['type'].unique()
print("Type takes on values: " + str(ins['type'].unique()) + '.')

Type takes on values: ['routine' 'complaint'].


In [35]:
ins['type'].nunique() #counting the types within using nunique

The number of occurences of each value in the DataFrame are 2.


In [40]:
print(ins[ins['type'] == 'routine'].shape[0])
print(ins[ins['type'] == 'complaint'].shape[0])

15429
1


In [43]:
print("The number of occurrences of routine are " + str(ins[ins['type'] == 'routine'].shape[0]) + '.')
print("The number of occurrences of complaint are " + str(ins[ins['type'] == 'complaint'].shape[0]) + '.')

The number of occurrences of routine are 15429.
The number of occurrences of complaint are 1.


In [22]:
ins2 = ins[ins['type']=='routine']
ins2

business_id  score      date     type
0               10     82  20160503  routine
1               10     94  20140729  routine
2               10     92  20140114  routine
3               19     94  20160513  routine
4               19     94  20141110  routine
...            ...    ...       ...      ...
15425        89515     98  20161103  routine
15426        89701    100  20161206  routine
15427        89790     96  20161129  routine
15428        89803    100  20161215  routine
15429        89900     93  20161206  routine

[15429 rows x 4 columns]

#### Question 5c

Since the data was stored in a .csv file, the dates are formatted as strings such as `20160503`. Once we read in the data, we would like to have dates in an appropriate format for analysis. Add a new column called `year` by capturing the first four characters of the date column. 

Hint: we have seen multiple ways of doing this in class, includings `str` operations, `lambda` functions, `datetime` operations, and others. Choose the method that works best for you :)

In [23]:
#first we have convert our date-like column to a datetime column
ins['date'] = pd.to_datetime(ins['date'], format ='%Y%m%d')
ins['year'] = ins['date'].dt.year
print(ins)

       business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
1               10     94 2014-07-29  routine  2014
2               10     92 2014-01-14  routine  2014
3               19     94 2016-05-13  routine  2016
4               19     94 2014-11-10  routine  2014
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[15430 rows x 5 columns]


#### Question 5d

What range of years is covered in this data set? Are there roughly same number of inspections each year? Try dropping records for any years with less than 50 inspections and store the result in a new dataframe named `ins3`.

In [24]:
min_year = ins['year'].min()
max_year = ins['year'].max()
print("Range of years covered in the dataset:", min_year, "-", max_year)

Range of years covered in the dataset: 2013 - 2016


In [25]:
#To count the number of inspections for each year
inspection_counts = ins['year'].value_counts().sort_index()
print("Number of inspections for each year:")
print(inspection_counts)


Number of inspections for each year:
year
2013      38
2014    5629
2015    3923
2016    5840
Name: count, dtype: int64


In [26]:
#option 1
ins3a = ins[ins['year'] > 2013]
print(ins3a)

       business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
1               10     94 2014-07-29  routine  2014
2               10     92 2014-01-14  routine  2014
3               19     94 2016-05-13  routine  2016
4               19     94 2014-11-10  routine  2014
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[15392 rows x 5 columns]


In [27]:
#Option 2
#filtering out records for years with less than 50 inspections
ins3 = ins.groupby('year').filter(lambda x: len(x) >= 50)
print(ins3)

       business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
1               10     94 2014-07-29  routine  2014
2               10     92 2014-01-14  routine  2014
3               19     94 2016-05-13  routine  2016
4               19     94 2014-11-10  routine  2014
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[15392 rows x 5 columns]


Let's examine only the inspections for one year: 2016. This puts businesses on a more equal footing because [inspection guidelines](https://www.sfdph.org/dph/eh/Food/Inspections.asp) generally refer to how many inspections should occur in a given year.

In [28]:
ins_2016 = ins[ins['year'] == 2016]
print(ins_2016)

       business_id  score       date     type  year
0               10     82 2016-05-03  routine  2016
3               19     94 2016-05-13  routine  2016
6               24     98 2016-10-05  routine  2016
7               24     96 2016-03-11  routine  2016
10              31     84 2016-08-16  routine  2016
...            ...    ...        ...      ...   ...
15425        89515     98 2016-11-03  routine  2016
15426        89701    100 2016-12-06  routine  2016
15427        89790     96 2016-11-29  routine  2016
15428        89803    100 2016-12-15  routine  2016
15429        89900     93 2016-12-06  routine  2016

[5840 rows x 5 columns]


### Question 6

#### Question 6a

Merge the business and 2016 inspections data, keeping all businesses regardless of whether they show up in the inspections file. Show the first several rows of the resulting dataframe.

In [29]:
pd.merge(bus, ins_2016, how="outer", on=["business_id"])

business_id                                name  \
0              10                    TIRAMISU KITCHEN   
1              19               NRGIZE LIFESTYLE CAFE   
2              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
3              24  OMNI S.F. HOTEL - 2ND FLOOR PANTRY   
4              31      NORMAN'S ICE CREAM AND FREEZES   
...           ...                                 ...   
7607        90234                ACCESS SFUSD THE ARC   
7608        90236                      PHO DONG HUONG   
7609        90237                    WIN'S RESTAURANT   
7610        90238                    USA POWER MARKET   
7611        90239                          ALBA RAY'S   

                            address           city state postal_code  \
0                     033 BELDEN PL  San Francisco    CA       94104   
1      1200 VAN NESS AVE, 3RD FLOOR  San Francisco    CA       94109   
2     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
3     500 CALIFORNIA ST, 2ND  FLOOR  San Francisco    CA       94104   
4              2801 LEAVENWORTH ST   San Francisco    CA       94133   
...                             ...            ...   ...         ...   
7607                1520 HOWARD ST   San Francisco    CA       94103   
7608             667 MONTEREY BLVD   San Francisco    CA       94127   
7609               3040 TARAVAL ST   San Francisco    CA       94116   
7610                1524 OCEAN AVE   San Francisco    CA       94112   
7611               2293 MISSION ST   San Francisco    CA       94110   

       latitude   longitude  phone_number  score       date     type    year  
0     37.791116 -122.403816  +14154217044   82.0 2016-05-03  routine  2016.0  
1     37.786848 -122.421547  +14157763262   94.0 2016-05-13  routine  2016.0  
2     37.792888 -122.403135  +14156779494   98.0 2016-10-05  routine  2016.0  
3     37.792888 -122.403135  +14156779494   96.0 2016-03-11  routine  2016.0  
4     37.807155 -122.419004           NaN   84.0 2016-08-16  routine  2016.0  
...         ...         ...           ...    ...        ...      ...     ...  
7607        NaN         NaN           NaN    NaN        NaT      NaN     NaN  
7608        NaN         NaN  +14154945048    NaN        NaT      NaN     NaN  
7609        NaN         NaN  +14158606036    NaN        NaT      NaN     NaN  
7610        NaN         NaN  +14153331878    NaN        NaT      NaN     NaN  
7611        NaN         NaN  +14158881393    NaN        NaT      NaN     NaN  

[7612 rows x 13 columns]

#### Question 6b
Print the 20 lowest rated businesses names, their addresses, and their ratings.

In [30]:
merged_df = pd.merge(bus, ins, how="outer", on=["business_id"])
sorted_df = merged_df.sort_values(['score'], ascending=True).head(20)
lowest_rated = sorted_df[['name', 'address', 'score']].head(20)
print(lowest_rated)

                                        name                     address  \
10699                        ABC BAKERY CAFE             650 JACKSON ST    
6385                         IMPERIAL PALACE          818 WASHINGTON ST    
9285          YUMMY DIM SUM & FAST FOOD, LLC            930 STOCKTON ST    
11745                GOLDEN RIVER RESTAURANT            5827 GEARY BLVD    
2453                             LES JOULINS                44 ELLIS ST    
2433                        YEE'S RESTAURANT             1131 GRANT AVE    
8655        OZONE THAI RESTAURANT AND LOUNGE                598 02ND ST    
2722             BASIL THAI RESTAURANT & BAR             1175 FOLSOM ST    
5929                           BODEGA BISTRO              607 LARKIN ST    
6733                     BRISTOL FARMS #2317           845 MARKET ST 010   
12113                        MISS SAIGON LLC                100 06TH ST    
7774                              FIRST CAKE            5901 GEARY BLVD    
15728       

## Done!

Now commit this notebook to your `hw3_submissions` branch, push it to your GitHub repo, and open a PR!